### ISSUE TITLE
[Bug] The current Starrocks is not using the correct dependencies version, and also can't build correctly.

### ISSUE THREAD

**Issue by matchyc** [StarRocks/starrocks#23111]
# TLDR; 
Users can't build Starrocks from the source. And Starrocks is now using incorrect dependencies which does not follow the restriction.

### Furthermore, even if we successfully pack the project for distribution, the several packages inside the libraries may come from the "packing machine", not the wanted versions of the library indicated in Starrocks' third-party restrictions.

# Steps to reproduce the behavior (Required)
git clone https://github.com/StarRocks/starrocks.git
cd starrocks
./build -j32 # just paralleled

# Expected behavior (Required)
BUILD SUCCESS

# Real behavior (Required)
Several Errors Occured

# StarRocks version (Required)
- branch main
- branch 3.0

# dive into codes & explain

> cmake version 3.26

Tow main parts:
1. third-party build errors - the procedure may not correct
2. one unused variable in starrocks

## third-party building
The error is mainly concentrated in the `apache arrow` build section.
Several packages can't be found by arrow even starrocks already built them before arrow.

### Error 1: CMake Error: Could NOT find Snappy (missing: Snappy_LIB)
The error below occurred even though we compiled snappy before building arrow.
```bash
CMake Error at /usr/local/bin/cmake/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find Snappy (missing: Snappy_LIB)
Call Stack (most recent call first):
  /usr/local/bin/cmake/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  cmake_modules/FindSnappy.cmake:55 (find_package_handle_standard_args)
  cmake_modules/ThirdpartyToolchain.cmake:235 (find_package)
  cmake_modules/ThirdpartyToolchain.cmake:948 (resolve_dependency)
  CMakeLists.txt:515 (include)
```
This is mainly caused by the wrong library type used in `build-thirdparty.sh`. In the standard operation, we should not install `libsnappy-dev` by `apt` in Ubuntu or `yum` in Centos, so there is something wrong when building.

According to the `CMakeLists.txt` which includes `ThirdpartyToolchain` in Line 515, `ThirdpartyToolchain.cmake` is the cmake module in the Apache arrow, which contains snappy and other dependencies needed by arrow 
>(some of them are built by `build-thirdparty.sh` before we build `arrow` such as `snappy` is built before `arrow` building, but some building procedures are still triggered by `arrow`).

So the snappy **should be found**. According to `FindSnappy.cmake` in `arrow`, I noticed that 
```cmake
if(ARROW_SNAPPY_USE_SHARED)
  set(SNAPPY_LIB_NAMES)
 # other things...
endif()
```
After checking this `ARROW_SNAPPY_USE_SHARED` setting, I found that in the current build settings, starrocks is trying to use DYNAMIC LIBRARY of Snappy to build arrow, but most of the packages are built to STATIC. So we will never find Snappy for arrow appropriately, because the library name we tried to find is **libsnappy.so**. 

How to fix: just use `-DARROW_SNAPPY_USE_SHARED=OFF` in `build-thirdparty.sh` when building arrow. Note that definitions like `-DARROW_BUILD_SHARED=OFF` or `-DARROW_BUILD_STATIC=ON` are not sufficient for arrow to use all static library for dependencies (already tested).

### Error 2: CMakeError: Count NOT find RapidJSONAlt

Error output:
```cmake
CMake Error at /usr/local/bin/cmake/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find RapidJSONAlt (missing: RAPIDJSON_INCLUDE_DIR) (Required is
  at least version "1.1.0")
Call Stack (most recent call first):
  /usr/local/bin/cmake/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  cmake_modules/FindRapidJSONAlt.cmake:69 (find_package_handle_standard_args)
  cmake_modules/ThirdpartyToolchain.cmake:235 (find_package)
  cmake_modules/ThirdpartyToolchain.cmake:1839 (resolve_dependency)
  CMakeLists.txt:515 (include)
```
Using the same thoughts to check the building procedure, this error is mainly caused by missing RapidJSON directory which should be provided to arrow cmake module.

**Note again**: we can manually install RapidJSON package by apt or yum in different systems, but if we do that, then the third-party section in starrocks will be meaningless.

**How to fix**: add cmake definition `-DRapidJSON_ROOT=$TP_INSTALL_DIR` in `build-thirdparty.sh` to help arrow find the `RapidJSON` package in the STARROCKS own thrid-party directory. 

Then we can get:
```
-- Found RapidJSONAlt: <my own directory>/starrocks/thirdparty/installed/include
```

### Error 3:  Could NOT find Lz4 (missing: LZ4_LIB LZ4_INCLUDE_DIR)

Same error type as before and the same inspecting method.

Add cmake options in `build-thirdparty.sh`. 
```
-DLZ4_INCLUDE_DIR=$TP_INSTALL_DIR/include/lz4 \
-DARROW_LZ4_USE_SHARED=OFF \
```

### Error 4: Not using zlib in third-party

Though `zlib.so` in my python virtual environment is found by arrow. But again, we need to use `zlib.a`.

**How to fix**: add cmake build definition `-DZLIB_ROOT=$TP_INSTALL_DIR` while building arrow. 

Note: this solution also meets a TODO task in arrow. 
>  TODO: Don't use global includes but rather target_include_directories

The more important thing is this enables Starrocks to use the correct & specific version of dependencies.

Then we can have:
```
Found ZLIB: /home/cm/projects/starrocks/thirdparty/installed/lib/libz.a (found version "1.2.11")  
```

### Error 5 & 6: Brotli is also using the library in the environment and files missing

Now arrow also finds brotli in my Python environment.
```
-- Found Brotli: <my miniconda path>/envs/py38/lib/libbrotlicommon.so
```
and files are missing
```
cp: cannot stat './brotli_ep/src/brotli_ep-install/lib/libbrotlienc-static.a': No such file or directory
```
Here is the code lines triggering the errors. [3 lines in total, 600-602] (https://github.com/StarRocks/starrocks/blob/cfea2ec3002254c93207523ff699491a6ace13af/thirdparty/build-thirdparty.sh#L600)

This is mainly because `brotli` will not be built by arrow. There is just non-exist the several files or directories we want to copy by command `cp`.

And more **IMPORTANT** is that I found that the library `libbrotlienc.a`, `libbrotlidec.a`, `libbrotlicommon.a` are needed by Starrocks main building procedure.
So if we don't install the library manually, we should fix the building error.

**How to fix**:
add `build_brotli()` function in `build-thirdparty.sh` script then call it to trigger building.
```
build_brotli() {
    check_if_source_exist $BROTLI_SOURCE
    cd $TP_SOURCE_DIR/$BROTLI_SOURCE
    mkdir -p $BUILD_DIR
    cd $BUILD_DIR
    rm -rf CMakeCache.txt CMakeFiles/
    ${CMAKE_CMD} .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=$TP_INSTALL_DIR
    ${CMAKE_CMD} --build . --config Release --target install

    cp -rf $TP_INSTALL_DIR/lib/libbrotlienc-static.a $TP_INSTALL_DIR/lib64/libbrotlienc.a
    cp -rf $TP_INSTALL_DIR/lib/libbrotlidec-static.a $TP_INSTALL_DIR/lib64/libbrotlidec.a
    cp -rf $TP_INSTALL_DIR/lib/libbrotlicommon-static.a $TP_INSTALL_DIR/lib64/libbrotlicommon.a
}
```

and call `build_brotli` before `build_arrow`

Also add cmake definitions
``` 
 -DBROTLI_ROOT=$TP_INSTALL_DIR \
-DARROW_BROTLI_USE_SHARED=OFF \
```

Then we have 
```
-- Found Brotli: <my home directory>/starrocks/thirdparty/installed/lib/libbrotlicommon-static.a 
```

### Error 7: Declared but not used variable

Due to Starrocks treating warnings as errors in cpp.

Here is a warning (also an error).

(https://github.com/StarRocks/starrocks/blob/cfea2ec3002254c93207523ff699491a6ace13af/be/src/exprs/agg/stream/retract_maxmin.h#L239)

`column` defined but never used. It seems that the if branch 
(https://github.com/StarRocks/starrocks/blob/cfea2ec3002254c93207523ff699491a6ace13af/be/src/exprs/agg/stream/retract_maxmin.h#L240)
is also meaningless. 
For passing build and having no effects for behaviors, I propose just deleting the several lines. Or a student like me who wants to contribute to Starrocks will never build the project successfully...

I would like to provide a PR to solve several questions!

**Comment by kevincai** [comment]
thanks for the feedback, will look into these issues one by one.
